src/components/install-app.tsx

```diff
@@ -0,0 +1,159 @@
+      };
+      // TODO: show some useful message incase some one loads this page from a
+      // non offline host, or in a browser without service worker support
+      return;
+    }
+
+    const offlineManifestId = queryValue("offlineManifest");
+    const loadingOfflineManifest = !!offlineManifestId;
+
+    this.state = {
+      serviceWorkerVersionInfo: "Starting...",
+      loadingOfflineManifest
+    };
+
+    this.wb = new Workbox("service-worker.js");
+
+    this.wb.addEventListener("waiting", (event) => {
+      // TODO: in future work we should show a dialog using this recipe:
+      // https://developers.google.com/web/tools/workbox/guides/advanced-recipes#offer_a_page_reload_for_users
+      // For now just send a message to skip waiting so we don't have to do it manually in the devtools
+      // This will trigger a 'controlling' event which we are listening for below and reload the page when
+      // we get it
+      this.wb.messageSkipWaiting();
+    });
+
+    navigator.serviceWorker.ready.then(registration => {
+      // Cache the application as soon as there is an active service worker
+      // it doesn't matter if the serivce worker is controlling the page because
+      // the fetches are done by the service worker not the page
+      this.cacheApplication();
+    });
+
+    this.wb.register().then((_registration) => {
+      console.log("Workbox register() promise resolved", _registration);
+
+      console.log("Sending GET_VERSION_INFO to service worker...");
+      this.setState({serviceWorkerVersionInfo: "Checking..."});
+      this.wb.messageSW({type: "GET_VERSION_INFO"})
+        .then((versionInfo: unknown) => {
+          if (typeof versionInfo === "string") {
+            this.setState({serviceWorkerVersionInfo: versionInfo});
+          } else {
+            this.setState({serviceWorkerVersionInfo: "Invalid response!"});
+          }
+        })
+        .catch(() => {
+          this.setState({serviceWorkerVersionInfo: "No response!"});
+        });
+    });
+
+    if (offlineManifestId) {
+      getOfflineManifest(offlineManifestId)
+        .then(async (offlineManifest) => {
+          console.log("got offlineManifest", offlineManifest);
+          if (offlineManifest) {
+            await saveOfflineManifestToOfflineActivities(offlineManifest);
+            this.setState({offlineManifest});
+          }
+        })
+        .catch(error => {
+          console.error(error);
+        });
+    }
+  }
+
+  cacheApplication() {
+    const wbManifest = (window as any).__wbManifest;
+    // const offlineManifest: OfflineManifest = { name: "Application Manifest", activities: [], cacheList: [] };
+
+    // FIXME: this caching is going to happen after the initial page load
+    // that means that the manifest.json and index.html will be cached by chrome as it parses the
+    // manifest.  The index.html request comes from Chrome's new installable dection that makes an offline request
```

Comment:
dection?

Reply:
I updated this comment, might have added more spelling mistakes in the process though. :)